odl/discr/discr_utils.py

```diff
@@ -817,6 +834,573 @@ def __init__(self, coord_vecs, values, input_type):
+    # wrapping. If needed, vectorize.
+
+    if callable(func_or_arr):
+        # Assume scalar float out dtype for single function
+        if out_dtype is None:
+            out_dtype = np.dtype('float64')
+
+        # Got a (single) function, possibly need to vectorize
+        func = func_or_arr
+
+        # Get default implementations if necessary
+        has_out, out_optional = _func_out_type(func)
+        if not has_out:
+            # Out-of-place-only
+            func_ip = partial(_default_ip, func)
+            func_oop = func
+        elif out_optional:
+            # Dual-use
+            func_ip = func_oop = func
+        else:
+            # In-place-only
+            func_ip = func
+            func_oop = partial(_default_oop, func)
+
+    else:
+        # This is for the case that an array-like of callables is provided.
+        # We need to convert this into a single function that returns an
+        # array.
+
+        arr = np.array(func_or_arr, dtype=object)
+
+        if val_shape is None:
+            # Infer value shape if `out_dtype is None`
+            val_shape = arr.shape
+        elif arr.shape != val_shape:
+            # Otherwise, check that the value shape matches the dtype shape
+            raise ValueError(
+                'invalid `func_or_arr` {!r}: expected `None`, a callable or '
+                'an array-like of callables whose shape matches '
+                '`out_dtype.shape` {}'.format(func_or_arr, val_shape)
+            )
+
+        out_dtype = np.dtype((scalar_out_dtype, val_shape))
+
+        arr = arr.ravel().tolist()
+
+        def array_wrapper_func(x, out=None, **kwargs):
+            """Function wrapping an array of callables and constants.
+
+            This wrapper does the following for out-of-place
+            evaluation (when ``out=None``):
+
+            1. Collect the results of all function evaluations into
+               a list, handling all kinds of sequence entries
+               (normal function, ufunc, constant, etc.).
+            2. Broadcast all results to the desired shape that is
+               determined by the space's ``out_shape`` and the
+               shape(s) of the input.
+            3. Form a big array containing the final result.
+
+            The in-place version is simpler because broadcasting
+            happens automatically when assigning to the components
+            of ``out``. Hence, we only have
+
+            1. Assign the result of the evaluation of the i-th
+               function to ``out_flat[i]``, possibly using the
+               ``out`` parameter of the function.
+            """
+            if is_valid_input_meshgrid(x, domain.ndim):
+                scalar_out_shape = out_shape_from_meshgrid(x)
+            elif is_valid_input_array(x, domain.ndim):
+                scalar_out_shape = out_shape_from_array(x)
+            else:
+                raise RuntimeError('bad input')
```

Comment:
Hm, yeah, you're right. I think there's an error message somewhere for the same situation. I'll try to dig it out.

Reply:
I think I could generalize [this function](https://github.com/odlgroup/odl/blob/b017445809567588f69828064ef816e5562c0229/odl/discr/discr_utils.py#L209-L249)

Reply:
Actually that won't be necessary since the check is always done [here](https://github.com/odlgroup/odl/blob/b017445809567588f69828064ef816e5562c0229/odl/discr/discr_utils.py#L1226-L1253). So at this point we should not get any error anymore, and that scenario would indeed be a `RuntimeError`.